### PR TITLE
[FEATURE] 3D identify tool working on 3D entities

### PR DESCRIPTION
**Pull request by wonder-sk**
Until now the tool only considered terrain. This commit adds support
for 3D renderers created from vector layers, so it is possible to
correctly identify polygons and linestrings in 3D.

In other words, you can pick a building extruded from its footprint or some other piece of a 3D geometry of a vector layer and it will magically work :-)

Founded by Politecnico di Torino, Dipartimento di Ingegneria dell'Ambiente, del Territorio e delle Infrastrutture (DIATI) within RESCULT project in collaboration with Faunalia

### CODE REVIEW

File: src/3d/qgs3dmapscene.h [hunk 1]
```diff
@@ -76,6 +78,11 @@ class _3D_EXPORT Qgs3DMapScene : public Qt3DCore::QEntity
     //! Returns the current state of the scene
     SceneState sceneState() const { return mSceneState; }
 
+    //! Registers an object that will get results of pick events on 3D entities. Does not take ownerhip of the pick handler. Adds object picker components to 3D entities.
```
Comment (nyalldawson) [inline src/3d/qgs3dmapscene.h]: "ownership"

File: src/3d/qgs3dmapscenepickhandler.h [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ * or pick handlers.
+ *
+ * This is used for identify tool to be able to identify entities coming from 3D renderers assigned to map layers.
+ *
+ * \since QGIS 3.4
+ */
+class Qgs3DMapScenePickHandler
+{
+  public:
+    virtual ~Qgs3DMapScenePickHandler() {}
```
Comment (nyalldawson) [inline src/3d/qgs3dmapscenepickhandler.h]: = default

File: src/3d/qgstessellatedpolygongeometry.h [hunk 3]
```diff
@@ -58,14 +60,20 @@ class QgsTessellatedPolygonGeometry : public Qt3DRender::QGeometry
     void setAddBackFaces( bool add ) { mAddBackFaces = add; }
 
     //! Initializes vertex buffer from given polygons. Takes ownership of passed polygon geometries
-    void setPolygons( const QList<QgsPolygon *> &polygons, const QgsPointXY &origin, float extrusionHeight, const QList<float> &extrusionHeightPerPolygon = QList<float>() );
+    void setPolygons( const QList<QgsPolygon *> &polygons, const QList<QgsFeatureId> &featureIds, const QgsPointXY &origin, float extrusionHeight, const QList<float> &extrusionHeightPerPolygon = QList<float>() );
+
+    //! Returns ID of the feature to which given triangle index belongs (used for picking)
+    QgsFeatureId triangleIndexToFeatureId( uint triangleIndex );
```
Comment (nyalldawson) [inline src/3d/qgstessellatedpolygongeometry.h]: const?

File: src/app/3d/qgs3dmaptoolidentify.cpp [hunk 4]
```diff
@@ -26,10 +26,42 @@
...
+
 Qgs3DMapToolIdentify::Qgs3DMapToolIdentify( Qgs3DMapCanvas *canvas )
   : Qgs3DMapTool( canvas )
 {
   connect( mCanvas->scene(), &Qgs3DMapScene::terrainEntityChanged, this, &Qgs3DMapToolIdentify::onTerrainEntityChanged );
+
+  mPickHandler.reset( new Qgs3DMapToolIdentifyPickHandler( this ) );
+}
+
+Qgs3DMapToolIdentify::~Qgs3DMapToolIdentify()
```
Comment (nyalldawson) [inline src/app/3d/qgs3dmaptoolidentify.cpp]: = default

**Review by nyalldawson — COMMENTED** [review]
Looks good! Any unit tests coming?

**Comment by wonder-sk** [comment]
> Any unit tests coming?

It is a bit difficult to properly unit test this as the ray casting is done internally by Qt3D. Ray casting API is available since Qt 5.11 so once this becomes minimum version for QGIS 3D I will be happy to have some more thorough auto testing...

**Pull request merged**